src/drivers/imx/esai.c

```diff
@@ -4,34 +4,387 @@
+			ESAI_xCR_xIE);
+	/* enable exception interrupt */
+	dai_update_bits(dai, REG_ESAI_xCR(direction), ESAI_xCR_xEIE,
+			ESAI_xCR_xEIE);
+	tracev_esai("ESAI_REGS_DUMP in esai_start");
+	esai_regs_dump(dai);
+}
+
+static void esai_stop(struct dai *dai, int direction)
+{
+	direction = direction == SOF_IPC_STREAM_PLAYBACK;
+	/* disable exception interrupt */
+	dai_update_bits(dai, REG_ESAI_xCR(direction), ESAI_xCR_xEIE, 0);
+	/* disable regular interrupt */
+	dai_update_bits(dai, REG_ESAI_xCR(direction), ESAI_xCR_xIE, 0);
+
+	dai_update_bits(dai, REG_ESAI_xCR(direction),
+			direction ? ESAI_xCR_TE_MASK : ESAI_xCR_RE_MASK, 0);
+
+	dai_update_bits(dai, REG_ESAI_xSMA(direction), ESAI_xSMA_xS_MASK, 0);
+	dai_update_bits(dai, REG_ESAI_xSMB(direction), ESAI_xSMB_xS_MASK, 0);
+
+	/* disable and reset FIFO */
+	dai_update_bits(dai, REG_ESAI_xFCR(direction),
+			ESAI_xFCR_xFR | ESAI_xFCR_xFEN, ESAI_xFCR_xFR);
+	dai_update_bits(dai, REG_ESAI_xFCR(direction), ESAI_xFCR_xFR, 0);
+	tracev_esai("ESAI_REGS_DUMP in esai_stop");
+	esai_regs_dump(dai);
+}
+
 static int esai_trigger(struct dai *dai, int cmd, int direction)
 {
+	tracev_esai("ESAI: trigger");
+	switch (direction) {
+	case DAI_DIR_PLAYBACK:
+		tracev_esai("ESAI: playback trigger");
+		break;
+	case DAI_DIR_CAPTURE:
+		tracev_esai("ESAI: capture trigger");
+		break;
+	default:
+		trace_esai_error("ESAI: <INVALID DIRECTION %d> trigger", direction);
+	}
+	switch (cmd) {
+#define CASE(c) case COMP_ ## c: tracev_esai("ESAI: trigger cmd DAI_" #c ); break;
+#define CASE2(c, p) case COMP_ ## c: tracev_esai("ESAI: trigger cmd DAI_" #p); break;
+	case COMP_TRIGGER_START:
+		tracev_esai("ESAI: trigger cmd DAI_TRIGGER_START");
+		esai_start(dai, direction);
+		break;
+	case COMP_TRIGGER_STOP:
+		tracev_esai("ESAI: trigger cmd DAI_TRIGGER_STOP");
+		esai_stop(dai, direction);
+		break;
+	case COMP_TRIGGER_PAUSE:
+		tracev_esai("ESAI: trigger cmd DAI_TRIGGER_PAUSE_PUSH");
+		esai_stop(dai, direction);
+		break;
+	case COMP_TRIGGER_RELEASE:
+		tracev_esai("ESAI: trigger cmd DAI_TRIGGER_PAUSE_RELEASE");
+		esai_start(dai, direction);
+		break;
+	/* Remaining triggers are no-ops, just print */
+	CASE(TRIGGER_SUSPEND)
+	CASE(TRIGGER_RESUME)
+#undef CASE
+	default:
+		trace_esai_error("ESAI: invalid trigger cmd %d", cmd);
+		break;
+	}
 	return 0;
 }
 
+static void esai_irq(void *ign)
+{
+	tracev_esai("ESAI: irq");
+}
+
 static int esai_probe(struct dai *dai)
 {
+	tracev_esai("ESAI: probe");
+	/* ESAI core reset */
+	dai_write(dai, REG_ESAI_ECR, ESAI_ECR_ERST | ESAI_ECR_ESAIEN);
+	/* We should do ESAI individual reset for the FIFOs */
+	dai_write(dai, REG_ESAI_TFCR, ESAI_xFCR_xFR);
+	dai_write(dai, REG_ESAI_RFCR, ESAI_xFCR_xFR);
+	/* Clear TSMA, TSMB */
+	dai_write(dai, REG_ESAI_TSMA, 0);
+	dai_write(dai, REG_ESAI_TSMB, 0);
+	dai_write(dai, REG_ESAI_RSMA, 0);
+	dai_write(dai, REG_ESAI_RSMB, 0);
+
+	dai_write(dai, REG_ESAI_ECR, ESAI_ECR_ESAIEN);
+
+	int irq = irqstr_get_sof_int(ESAI_IRQ);
```

Comment:
Can you declare the variables at the start of the function.

Reply:
Will do wherever I've done this (I think Checkpatch knows how to complain about that). This particular instance may be resolved differently though (the irq computation may be moved in get_handshake)

Reply:
For the moment I did what you said.